rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,484 @@
+  Running,   // The session has at least one running search ID associated with it.
+  Done,      // All search IDs associated with this session have completed.
+  Error,     // At least one search ID associated with this session had an error. 
+  Expired,   // The session has expired. Associated search ID data was cleared from ES.
+}
+```
+
+### Saved Object Structure
+
+```ts
+interface BackgroundSessionAttributes extends SavedObjectAttributes {
+  sessionId: string;
+  status: BackgroundSessionStatus;
+  name: string;
+  creation: string;
+  expiration: string;
+  idMapping: { [key: string]: string };
+  url: string; // TODO
+  metadata: { [key: string]: any } // TODO
+}
+```
+
+## Frontend Services
+
+Most sessions will probably not be saved. Therefore, to avoid creating unnecessary saved objects, the browser will keep track of requests and their respective search IDs, until the user chooses to store the session. Once a session is stored, any additional searches will be immediately saved on the server side. 
+
+### New Session Service
+
+We will expose a new frontend `session` service on the `data` plugin `search` service. 
+
+The service will expose the following APIs:
+
+```ts
+interface ISessionService {
+   /** 
+    * Returns the current session ID
+    */
+   getId: () => string;
+
+   /**
+    * Start a new session, by generating a new session ID.
+    */
+   start: () => string;
+
+   /**
+    * Restore a session. Any searches executed using this ID will attempt to reload saved data.
+    * A restored session may be expired, meaning that following search requests will fail. 
+    * @param sessionId Session ID to restore.`.
+    */
+   restore: (sessionId: string) => void;
+
+   /**
+    * Close the current session and start a new one
+    */
+   clear: () => void; 
+
+   /**
+    * Store a session.
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @param name A display name for the session.
+    * @param url TODO: is the URL provided here? How?
+    * @param searchIdMapping An optional mapping of hashed requests mapped to the corresponding searchId.
+    * @returns The stored `BackgroundSessionAttributes` object
+    * @throws Throws an error in OSS.
+    */
+   store: (sessionId: string, name: string, url: string, searchIdMapping?: Record<string, string>)) => Promise<BackgroundSessionAttributes>
+
+   /**
+    * @returns Is the current session stored (i.e. is there a saved object correspondng with this sessionId).
+    */
+   isStored: () => boolean;
+
+   /** 
+    * @returns Is the current session restored (i.e. was loaded by calling `sessionService.restore()`).
+    */
+   isRestore: () => boolean;
+
+   /**
+    * Mark a session as and all associated searchIds as expired.
+    * Cancels active requests, if there are any.
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @returns success status
+    * @throws Throws an error in OSS.
```

Comment:
We will need to make sure any places that call these methods have an appropriate `catch` block to handle when these methods are unavailable. This might be useful for showing some sort of "upgrade to a free license" pop-up or some other call to action.

Reply:
Will there be a way to distinguish between a not-implemented in OSS error versus other error conditions? Personally, I dislike APIs which throw errors that should always be caught, especially in languages like Javascript/Typescript which are lacking equivalent features to Java's checked exceptions.

Reply:
On the other hand, if an API don't return any response (like `sessionService.restore(sessionId)` for example), we won't know it failed.
What would be an alternative?